ompi/mca/mtl/ofi/mtl_ofi_component.c

```diff
@@ -736,6 +749,11 @@ ompi_mtl_ofi_component_init(bool enable_progress_threads,
                         "%s:%d: fi_getinfo(): %s\n",
                         __FILE__, __LINE__, fi_strerror(-ret));
 
+    if (unset_fi_hmem_cuda_enable_xfer) {
```

Comment:
It looks like you're taking the strategy of:

* Set the env var
* Call `fi_getinfo()`
* Unset the env var

What are the consequences / interactions with other places in the Open MPI code base that also call `fi_getinfo()` -- is it ok to temporarily set this libfabric behavior and then unset it?

Reply:
Good question - on Libfabric's end, `FI_HMEM_CUDA_ENABLE_XFER` is checked during its own HMEM initialization, which is ultimately triggered by either `fi_getinfo()` or `fi_fabric()`. Internally, `FI_HMEM_CUDA_ENABLE_XFER` is initialized once - effectively a `static` variable. So while HMEM initialization can be inadvertently triggered multiple times, multiple/subsequent calls to `fi_getinfo()`/`fi_fabric()` won't overwrite Libfabric's internal state for `FI_HMEM_CUDA_ENABLE_XFER`; unsetting the env var immediately after calling `fi_getinfo()` will have no effect in Libfabric.

So the idea here was to "wrap" the `fi_getinfo()` calls for the OFI component initialization. In my testing, wrapping the `fi_getinfo()` in the MTL OFI compeonent was sufficient, but I also "wrapped" the calls in the BTL OFI component and BTL usnic component per William's suggestion. I left the `fi_getinfo()` call alone in the BTL usnic module since it doesn't look like it's being called to initialize Libfabric.

Reply:
Gotcha.

If libfabric's internal state of this env variable is only initialized once, does this mean that the hmem CUDA support will only be enabled if one of your calls (that setenv's `FI_HMEM_CUDA_ENABLE_XFER` beforehand) happens to be the very first call to `fi_getinfo()` / `fi_fabric()`?

Reply:
`<laughing a bit>` We just spent several years blasting a particular device vendor for making that exact architectural choice - basing library behavior on a one-time view of an envar that multiple layers above it could be setting. Bit disappointing to see libfabric making the same error. After all, you don't have any way of knowing what the app or some other lib called by the app has done before getting into OMPI. So this becomes completely indeterministic.

Reply:
> does this mean that the hmem CUDA support will only be enabled if one of your calls [...] happens to be the very first call to `fi_getinfo()` / `fi_fabric()`?

@jsquyres The new Libfabric spec is each provider chooses whether to perform CUDA memory transfers when `FI_HMEM_CUDA_ENABLE_XFER` is unspecified. So Libfabric will initialize the interface unconditionally as long as it was compiled with CUDA support, but each provider has the freedom to choose its own default. For `FI_HMEM_CUDA_ENABLE_XFER`, Libfabric continues to check if it's set in the environment until it's explicitly specified. So you don't need to ensure you're making the very first call to `fi_getinfo()` or `fi_fabric()`; just the first call where `FI_HMEM_CUDA_ENABLE_XFER` is set (or let providers choose what to do).

For EFA specifically, the motivation here was to prevent CUDA memory operations where deadlocks or performance degradation may occur (e.g. NCCL). Hence, the EFA provider will only utilize CUDA memory transfers if `FI_HMEM_CUDA_ENABLE_XFER` is explicitly set.

Reply:
> basing library behavior on a one-time view of an envar that multiple layers above it could be setting.

@rhc54 Pedantically, it's not quite a "one-time" view; you don't need to set it before the *very first* `fi_getinfo()`/`fi_fabric()` call, but it *is* immutable once it's been specified.

Part of the overall context is that we (EFA) needed some mechanism to prevent unexpected behavior with older (non-OMPI) applications, while also allowing MPI applications to utilize CUDA memory transfers in Libfabric if CUDA buffers are passed. It turned out that the env var `FI_HMEM_CUDA_ENABLE_XFER` already existed (though virtually unused) - any other mechanism (env var, command-line flag, etc) would ostensibly conflict with `FI_HMEM_CUDA_ENABLE_XFER`. Since it previously defaulted to true, our only feasible choice was to reimplement `FI_HMEM_CUDA_ENABLE_XFER` as a ternary enum (`[UNSPEC, ENABLED, DISABLED]`). By virtue of the amount of time spent wrangling this issue, I can't necessarily disagree with your overall assessment/sentiment, though. I'm not exactly going to hang this one on my fridge.

Re: the mutability of `FI_HMEM_CUDA_ENABLE_XFER`, I'm not sure we can support some application/layer switching it on/off once it's been set; at least not in the current state. Off the top of my head, I know we (EFA) have some endpoint initialization logic that assumes `FI_HMEM_CUDA_ENABLE_XFER` is fixed. It likely boils down to implementation details, so perhaps it can be a future enhancement.

Reply:
So here is the problem. I am a higher layer (maybe the app itself) and I don't set the envar because I want the default behavior and don't realize that someone underneath me might undermine my decision. So now I am using the OFI provider in the belief that it is behaving one way, completely unaware that you changed it without my knowledge.

I don't think that sounds like a particularly user-friendly approach. However, they are _your_ customers, and if you don't mind getting yelled at when they discover you did this, then that's fine by me! Our prior experience resulted in exactly that problem - totally unexpected behavior because somebody changed the flag without an upper level knowing it happened...and people were very angry about it.

Even trying to warn people that you are doing this isn't all that helpful. First, not everyone will hear about it, and second existing codes won't necessarily know to go back and change it.

Taking the time to fix it right might be a better alternative. Our prior experience of spending years fighting to get it corrected was far from pleasant and left a bad taste in customer's mouths, even after we got it fixed.

Just a word of warning from someone who went thru this before - I'll leave it up to you from here.

Reply:
Was asked to provide a couple of our "horror stories" to help illustrate the problem:

* user specified the desired setting in the environment prior to starting the app, and then relied on it in their own use of the library. Unfortunately, a library they were using (proprietary and invisible to them) overwrote the envar because they could only use the library in one mode, and so they set it to what they needed. Spent months of hair tearing trying to figure out why nothing worked as expected as there was no way to know what happened, or when it happened.
* user has an app that is working fine, and then extends it by using some functionality from a standard library. If they init that library at one place in the code, things behave a certain way - move that init line to a slightly different place in the code, and suddenly the app behaves completely differently. Turned out each library was overwriting the envar to be what they needed, and so ordering of the library inits - even though the libraries were totally independent of each other - could cause unexpected behavior in the app.

I could go on - my list is rather long and impressive - but enough said. I hope it has provided enough "voice from painful experience" to help guide your design, and apologize if it feels like I've been harassing you. Not my intent at all.

Reply:
The crux of the issue is the use of environment variables. In a vacuum, it is absolutely a poor design choice to have implementation-level logic based on a mechanism any layer in the stack can manipulate, as opposed to something more conventional/principled like a capability bit in the `flags` passed to `fi_getinfo()`. Unfortunately, this is the hand we were dealt. I think you would agree it would be even worse to use a more principled approach like a capability bit while completely undermining any existing `FI_HMEM_CUDA_ENABLE_XFER` in the environment. At best, it's a less-than-ideal design, but I don't really see a clear path to avoid using `FI_HMEM_CUDA_ENABLE_XFER`. We have to respect it, but we also don't really want MPI users having to set `FI_HMEM_CUDA_ENABLE_XFER=1` for their CUDA-aware MPI workloads.

Re: your provided anecdotes: At least for this PR/OMPI, we are only setting the env var if it doesn't exist in the environment. Naturally we only have so much control/influence over each application layer, and it's obviously not difficult to manipulate env vars, so this could potentially become an issue in the future.